src/hooks/user/usePermission.ts

```diff
@@ -0,0 +1,95 @@
+import { useMemo } from "react";
+
+import { ALWAYS_EMPTY_ARRAY, COLLECTION_ROLES } from "settings";
+
+import { SpaceId, UserId, WorldId } from "types/id";
+
+import { useLiveDocument } from "hooks/fire/useLiveDocument";
+import { useSpacesByOwner } from "hooks/spaces/useSpacesByOwner";
+import { useWorldsByOwner } from "hooks/worlds/useWorldsByOwner";
+
+type AdminRole = {
+  allowAll: boolean;
+  users: string[];
+};
+
+type UsePermissionOptions = {
+  userId: UserId;
+  worldId?: WorldId;
+  spaceId?: SpaceId;
+};
+
+export const usePermission = (options: UsePermissionOptions) => {
+  const path = useMemo(() => [COLLECTION_ROLES, "admin"], []);
+
+  const {
+    data: role,
+    error: roleError,
+    isLoading: roleLoading,
+  } = useLiveDocument<AdminRole>(path);
+
+  const {
+    ownWorlds,
+    error: worldsError,
+    isLoading: worldLoading,
+  } = useWorldsByOwner(options);
+
+  const {
+    ownSpaces,
+    error: spacesError,
+    isLoading: spacesLoading,
+  } = useSpacesByOwner(options);
+
+  const { userId, worldId, spaceId } = options;
+
+  return useMemo(() => {
+    const isAnyWorldOwner = !worldLoading && ownWorlds?.length > 0;
+    const isAnySpaceOwner = !spacesLoading && ownSpaces?.length > 0;
+
+    const isWorldOwner =
+      !worldLoading && ownWorlds?.find(({ id }) => id === worldId);
+    const isSpaceOwner =
+      !spacesLoading && ownSpaces?.find(({ id }) => id === spaceId);
+
+    const users = role?.users;
+    const ids = !roleLoading && users ? role?.users : ALWAYS_EMPTY_ARRAY;
+    const isSuperAdmin = userId && ids.includes(userId);
+
+    const isLoading = worldLoading || spacesLoading || roleLoading;
+    const isLoaded = !isLoading;
+
+    const isAdmin = isSuperAdmin || isAnyWorldOwner || isAnySpaceOwner;
+    const isNotAdmin = !isAdmin;
+
+    const error = roleError ?? worldsError ?? spacesError;
```

Comment:
I worry about the performance impact of this function. In particular, if we call this twice with the same arguments from two different places do we end up doing the same queries against the database multiple times? Do we end up with multiple subscriptions for update to those bits of data?

Can you do some checking to see exactly what happens in terms of queries hitting the database? It might make sense to move some of these checks into a top level permissions context provider that does most of the work once and once only.

Reply:
> Do we end up with multiple subscriptions for update to those bits of data?
Yes. There is no project wide caching involved. For that I can suggest a follow up PR that adds a context that uses this hook, maybe even straight as part of those `LoginRestricted` and `WithPermission` components.

> Can you do some checking to see exactly what happens in terms of queries hitting the database? 

Short answer: can if you provide with a baseline. It's the same SDK be it used through one wrapper or another. the `useLive*` hooks are just more simplistic in design.

The long answer: 
it's not like a "number of queries hitting the database", 
but a regular protobuf channel that works the same for the API be it react-redux-connect or these `onSnapshot` listeners.

The "database driver" i.e. Firestore SDK has its own `*FromCache` and `*FromServer` versions that it decides which to use when a bare call like `getDoc` is made and the communication (you can notice in the browser's network tab) is made up of packaged deltas for the various connections and queries it has.

The listener itself already employs local caching and options to read only deltas or metadata for local changes that still are to be persisted to server etc.
https://firebase.google.com/docs/firestore/query-data/listen

Here is an excerpt from that documentation page:
>The initial state can come from the server directly, or from a local cache. If there is state available in a local cache, the query snapshot will be initially populated with the cached data, then updated with the server's data when the client has caught up with the server's state.